[CK_TILE] Stream-K Tree Reduction and Cache Skipping Integration

ecamartins opened the pull request:
## Proposed changes

This PR makes 2 main changes. A high-level description of the changes is as follows:
1. **Adds a Tree Reduction strategy to CK Tile Stream-K**:
In the original Stream-K reduction (implemented in [PR3107](https://github.com/ROCm/composable_kernel/pull/3107)), when there are N workgroups contributing to the same C tile, the first workgroup of the tile is responsible for reading and accumulating all partial results of the other N-1 workgroups from a global buffer. This takes O(N) steps, which can become a bottleneck when N is large. This PR introduces a new reduction strategy that avoids placing the burden on a single workgroup to accumulate all results for a C tile. This new strategy is a Tree Reduction, where all workgroups contributing to a C tile cooperatively accumulate results in parallel, reducing the number of steps for the reduction from O(N) to O(logN).
2. **Removes the use of atomics for signaling between workgroups, skipping cache instead**:
In the original non-atomic reduction strategy (implemented in [PR3107](https://github.com/ROCm/composable_kernel/pull/3107)), atomics were used to implement signaling between workgroups. The implementation used CK Tile's `workgroup_barrier`; however, this class uses a relaxed atomic memory ordering. Through experience implementing the tree reduction, it was discovered that the relaxed memory ordering is not enough to guarantee that workgroups will not read stale data from the global buffer shared between workgroups. Although stronger atomic memory orderings exist (e.g., acquire/release), these memory orderings are very expensive. Thus, this change leverages cache coherency modifiers to skip cache, thereby avoiding the use of atomics.

Some more specific changes are as follows:
- Use scalar loads and scalar stores with GLC flag to read and write to the flags array. This ensures that flags results are visible to other workgroups, skipping appropriate cache levels (see ISAs for more details). **Note:** This must be done via inline ASM as there are no builtins for scalar loads and stores at this time.
- Use either GLC and SLC (for gfx90a) or sc0 and sc1 (for gfx942 and gfx950) flags to write to partials. This ensures that partial results are written to a location visible to other workgroups, skipping appropriate cache levels (see ISAs for more details).
- Addition of unit tests for normal reduction and tree reduction (more will be added soon via tile engine)
- Remove the use of `workgroup_barrier`
- Moved stream-k kernel files to a subdirectory here: `include/ck_tile/ops/gemm/kernel/streamk_gemm`

**Note:** gfx908 support is not included in this change. More investigation is required to sort out order cache coherency flags. This will come in a future change.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

ecamartins committed:
CK Tile Stream-K Tree Reduction

This change adds the first implementation of the Stream-K tree reduction
strategy into CK Tile. The tree reduction reduces the the number of
steps for accumulating results for a tile from O(N) to O(logN) where N
is the number of workgroups contributing to a C tile.

Additionally, in the original non-atomic reduction strategy, atomics
were used to set the flags buffer and to read from the flags buffer.
Howeover, through investigation with the tree reduciton, atomics with
default (relaxed) semantics were not enough to guarantee workgroups
would not read stale data, leading to incorrect results. Stronger
acquire/release memory orderings are too expensive. So, this change
also eliminates the use of atomics for setting the flags. Instead, we
leverage cache modifiers (e.g., GLC) to avoid writing to cache, thereby
avoiding the use of atomics.

Prelimiary tests were also added for the normal reduction and tree
reduction. More will be added in a future PR via tile engine.

ecamartins committed:
Move Stream-K kernel files to a subdirectory

ecamartins commented on include/ck_tile/core/arch/amd_buffer_addressing.hpp:
```diff
@@ -1417,6 +1417,7 @@ enum struct amd_buffer_coherence_enum
     DEVICE_NT1 = 10,
     SYSTEM_NT0 = 9,
     SYSTEM_NT1 = 11,
+    sc0_sc1    = 17
```
This enum uses inconsistent casing:
<img width="713" height="411" alt="image" src="https://github.com/user-attachments/assets/4ac46b81-b824-4b37-98f7-1b07f2abdc84" />

I will make a separate ticket to clean up the casing style.

ecamartins commented on include/ck_tile/core/arch/amd_buffer_addressing_builtins.hpp:
```diff
@@ -1285,6 +1285,7 @@ enum struct amd_buffer_coherence_enum
     DEVICE_NT1 = 10,
     SYSTEM_NT0 = 9,
     SYSTEM_NT1 = 11,
+    sc0_sc1    = 17
```
I will make a separate ticket to unify the casing style.

cgmillette commented on include/ck_tile/ops/common/streamk_common.hpp:
```diff
@@ -8,7 +8,8 @@
 {
-    Atomic    = 0u,
-    Reduction = 1u
+    Atomic        = 0u,
```
Can you add this to your unify casing style ticket?
UPPER_SNAKE for these

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm/streamk_gemm_coherency.hpp:
```diff
@@ -0,0 +1,34 @@
+};
+
+template <typename CompilerTarget>
+struct StreamKCoherency<CompilerTarget,
```
Looks good!

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm/streamk_gemm_kernel.hpp:
```diff
@@ -318,37 +319,58 @@ struct StreamKKernel
+        auto sk_flags_ptr = static_cast<index_t*>(kargs.workspace_ptr);
+        index_t offset    = cta_idx * sizeof(index_t);
+
+        asm volatile("s_mov_b32 m0, %2\n\t"
```
Let's make a note to ask the compiler team for a scalar __builtin so we can give it coherence flags!

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm/streamk_gemm_kernel.hpp:
```diff
@@ -318,37 +319,58 @@ struct StreamKKernel
-     * CTA index.
+     * @note This function utilizes a scalar store to write to the flags buffer.
      */
     CK_TILE_DEVICE void SignalStorePartialDone(const StreamKKernelArgs& kargs,
```
For the casing ticket lets fix this to camelCase for function names

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm/streamk_gemm_kernel.hpp:
```diff
@@ -318,37 +319,58 @@ struct StreamKKernel
-        auto sk_flags_ptr = static_cast<uint32_t*>(kargs.workspace_ptr);
-        workgroup_barrier sk_flags(sk_flags_ptr);
-        sk_flags.wait_eq(1, cta_idx);
+        auto sk_flags_ptr = static_cast<index_t*>(kargs.workspace_ptr);
```
Just to help with context, we should write this as auto* sk_flags_ptr

ecamartins replied:
Done.

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm/streamk_gemm_kernel.hpp:
```diff
@@ -528,44 +556,100 @@ struct StreamKKernel
+                    }
                 }
-                else
+                else // Tree Reduction
```
I think it might be best to be explicit about a check for StreamKReductionStrategy::TreeReduction, just due to if the enum were to get more values later on this might lead to undesired behavior. The else clause should maybe throw an error because it's not a supported enum value.

ecamartins replied:
I agree. Since we can't throw in device functions, I used a static assert instead. I also moved the check to the outermost `if` block so it's easier to read. So, the line you commented on above will stay the same since the outer `if else` checks for `Reduction || Tree Reduction`.

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm/streamk_gemm_tile_partitioner.hpp:
```diff
@@ -46,6 +46,16 @@ struct StreamKTilePartitionerBase
+     * non-persistent DP section is used, then a Stream-K workgroup's `cta_idx` should be something
+     * like `blockIdx.x` minus number of DP workgroups.
+     */
+    CK_TILE_DEVICE index_t get_start_iter(index_t cta_idx) const noexcept;
```
Adding function naming to camelCase TBD ticket

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm/streamk_gemm_tile_partitioner_impl.hpp:
```diff
@@ -61,13 +61,21 @@ StreamKTilePartitionerBase<BlockGemmShapeType, ReductionStrategyType>::get_flags
+StreamKTilePartitionerBase<BlockGemmShapeType, ReductionStrategyType>::get_start_iter(
+    index_t cta_idx) const noexcept
+{
+    index_t extra_iters_before_me = ck_tile::min(cta_idx, extra_iters_);
```
Quick description of what extra_iters_before_me means in a comment would be nice

ecamartins replied:
Done :)

cgmillette commented on test/ck_tile/gemm_streamk/test_streamk_tile_partitioner.cpp:
```diff
@@ -372,6 +372,85 @@ TEST(StreamKTilePartitionerBaseGetOutputTileIndex, TestAllMappings)
+TEST(StreamKTilePartitionerBaseGetTileLocalCtaIndex, SKOnlyLargeK)
+{
+    /*
+    The StreamKTilePartitionerBaseConfigSKOnlyLargeK has the following form:
```
Really like the descriptions here!

cgmillette reviewed (commented):
So far, looks great just a few changes needed and I'll approve

ecamartins committed:
Cleanup Code Style & Handle Unsupported Reductions

This change makes the following small changes:
- Add an explicit else block for unimplemented reduction strategies
- Clarify type of sk_flags_ptr via auto*
- Add description for extra_iters_before_me variable

cgmillette reviewed (dismissed):
LGTM

ecamartins committed:
Run new copyright script on new files
